Reintroduce five-minute and fifteen-minute load averages on Linux

jasontedor opened the pull request:
This commit reintroduces the five-minute and fifteen-minute load stats
on Linux, and changes the format of the load_average field back to an
array.

Relates #12049, relates #14741

nik9000 commented:
Makes sense to me though someone else should look at it too.

rmuir commented:
Do you think its cleaner to remove some `#ifdef LINUX`/security code here and just unconditionally add a FilePermission read for `/proc/uptime` to `security.policy`?

Its not really going to do any harm on other OSes, and I think it makes it easier as a static permission this way (e.g., to factor out this monitoring stuff to a module with these permissions and more isolation, monitoring stuff is a great candidate for that). Having it in this file makes that refactoring more clear, because someone just has to move it to the `plugin-security.policy` file when they try to do that.

In general, we add most filesystem permissions dynamically with java code: not because we want to, but because its our only choice since the paths are configurable in e.g. elasticsearch.yml... but I always prefer the policy files if there is any way, even if its a little less secure! In this case the path is well-known/not-configurable, so that is the way I would do it.

rmuir commented:
I would remove the constant too, in this case it just causes a loss of locality. Think if i want to add freebsd support here, I see it like this:

```
security.policy:
// load average statistics on linux
grant FilePermission "/proc/loadavg", "read";
// load average statistics on freebsd
grant FilePermission "/compat/linux/proc/loadavg", "read";
```

Then we only have system-dependent code in one place: the monitoring code itself... makes it easier to move out :) and I think we would prefer to keep Environment simpler and limited to "core" stuff.

jasontedor commented:
> Do you think its cleaner to remove some #ifdef LINUX/security code here and just unconditionally add a FilePermission read for /proc/loadavg to security.policy?

@rmuir I do think it's cleaner, but I was trying to following the principle of least privilege any only grant those permissions on systems where it will actually be used. I pushed f7c7f8eb5231f7f317691a8a30f4594191730458 to make this permission static and defined in security.policy.

jasontedor commented:
> I would remove the constant too, in this case it just causes a loss of locality.

@rmuir I moved this constant to a new class to hold these in f7c7f8eb5231f7f317691a8a30f4594191730458. I added new class to holding these because more are coming (e.g., `/proc/[pid]/io`).

rmuir commented:
> @rmuir I do think it's cleaner, but I was trying to following the principle of least privilege any only grant those permissions on systems where it will actually be used. I pushed f7c7f8e to make this permission static and defined in security.policy.

Yeah, I see the rationale. Its a tough tradeoff, but in cases like this I think simplicity should win. Otherwise monitoring code could get too complicated to factor out and we miss the forest for the trees, because if we can pull it out into a module, then a ton of stuff including this (like messing around with arbitrary threads, environment variable access, sysprops, ...) can be moved to just that module: so only the 1% code that needs it has it.

jasontedor commented:
> I do think it's cleaner, but I was trying to following the principle of least privilege any only grant those permissions on systems where it will actually be used.

@rmuir Hmm...now I'm thinking about `/proc/[pid]/io` and wondering if we only want to grant this for the Elasticsearch pid which requires a dynamic setup of the property?

rmuir commented:
> @rmuir Hmm...now I'm thinking about /proc/[pid]/io and wondering if we only want to grant this for the Elasticsearch pid which requires a dynamic setup of the property?

that is why there is /proc/self

jasontedor commented:
> that is why there is /proc/self

@rmuir :)

rmuir commented:
> I added new class to holding these because more are coming (e.g., /proc/[pid]/io).

Well, see my freebsd example: in this case we have a totally different "de-facto" path (/compat/linux/...), and it asks you to set this up when you install openjdk (not sure if it will fail hard without it). Anyway I just don't think these constants being in a separate file helps much. To me its too much indirection.

I guess to me this is more intuitive, since it prevents system-dependent stuff from being sprayed everywhere:
```
if (Constants.Linux) {
   processLoadAverage("/proc/loadavg");
} else if (Constants.FreeBSD) {
   processLoadAverage("/compat/linux/proc/loadavg");
} ...
```

What does the separate file with constants buy us?

jasontedor commented:
> What does the separate file with constants buy us?

@rmuir What it buys is not spraying these paths in a bunch of different places which lets us use `@SupressForbidden` in one place instead of many.

rmuir commented:
But we don't truly need a separate file for that right? If we go with my pseudocode we have:
```
@SuppressForbidden(reason = "access absolute paths for operating system statistics")
private void something() {
  if (Constants.Linux) {
   processLoadAverage(PathUtils.get("/proc/loadavg"));
  } else if (Constants.FreeBSD) {
   processLoadAverage(PathUtils.get("/compat/linux/proc/loadavg"));
  } ...
}
```

I think this is just as good, and `processLoadAverage(Path loadFile)` with the try/catch and parsing logic wouldn't need any suppression, its agnostic to where the file is coming from. If its a static method it also makes unit testing it very easy (just make fake files).

jasontedor commented:
> But we don't truly need a separate file for that right?

@rmuir Nope. I pushed 67d1785cbb06ba0944960c30ed61e298f8e42723 and think this is ready for a final review?

rmuir commented on core/src/main/java/org/elasticsearch/monitor/os/OsStats.java:
```diff
@@ -166,22 +170,70 @@ public static Cpu readCpu(StreamInput in) throws IOException {
+            private double fiveMinuteLoadAverage;
+            private double fifteenMinuteLoadAverage;
+
+            public final static LoadAverage DEFAULT_LOAD_AVERAGE = new LoadAverage(-1, -1, -1);
```
does this constant make sense to be here or in the fallback code should we just pass `new LoadAverage(-1, -1, -1)`. Maybe we should remove the ctor taking a single `double` as well, and pass `new LoadAverage(x, -1, -1)`. I had trouble following the logic but then it would be all clear what values are being returned in those cases.

jasontedor replied:
> does this constant make sense to be here or in the fallback code should we just pass new LoadAverage(-1, -1, -1).

@rmuir I was looking to avoid creating unnecessary garbage when there is periodic monitoring in place and these statistics aren't available (specifically Windows).

> Maybe we should remove the ctor taking a single double as well, and pass new LoadAverage(x, -1, -1). I had trouble following the logic but then it would be all clear what values are being returned in those cases.

That's fine. I pushed

rmuir replied:
Also, honestly I am curious about why LoadAverage.java needs to exist. I would use a `double[]` representation, consistent with the JDK? I don't think it will be confusing, if someone is dealing with load average they expect this?

jasontedor replied:
> Also, honestly I am curious about why LoadAverage.java needs to exist.

@rmuir It's not needed, and it's how I originally coded it locally before thinking it was clearer like this.

rmuir replied:
We wouldn't have the confusing Double (it immediately makes me think, how/why can this be null). I expect we'd use -1 for the back compat case anyway? (e.g. just read a single double, and return { x, -1, -1 }).

jasontedor replied:
> We wouldn't have the confusing Double

Oh, that's just a mistake; it should have been a `double` with a lowercase '`d`'.

jasontedor replied:
@rmuir I removed `OsStats.Cpu.LoadAverage` and returned to using a `double[]` in 10863db685d52cd9d7e4c3217f739a1051d158f8. Personally I think that the dedicated class is simpler, but I have no issues keeping it this way if you feel otherwise.

rmuir commented:
I do think its a lot better. So the last thing is this `DEFAULT_LOAD_AVERAGE`. I'm really not a fan of this..., its static, but there is nothing to enforce immutability of the thing (regardless of double or pojo representation), so besides being a dangerous optimization, I'm not sure that optimizing this actually matters. 

If we really want to keep this optimization can it be at least moved to the method where its returned from, have a better name, and be private, like `private static final double[] LOAD_AVERAGE_UNAVAILABLE` ?

jasontedor commented:
@rmuir I pushed 5fb5ed8f39c1c617d4fe91f7d8c690ad657c2528 to just allocate new instances when the load averages are not available.

rmuir commented on core/src/main/java/org/elasticsearch/monitor/os/OsStats.java:
```diff
@@ -152,8 +156,9 @@ public void writeTo(StreamOutput out) throws IOException {
+
         short percent = -1;
-        double loadAverage = -1;
+        double[] loadAverage = new double[] { -1, -1, -1 };
```
what do we need to initialize this to a strange value vs `null`? Is `null` not good enough? I see confusing code like this in all these `Streamable`s.... to me it seems like the whole entire Streamable API is broken. I dont understand why we have readFrom() at all, vs. say using an abstract class with a ctor that takes Streamable, and thats just how things are deserialized. Then these various classes could be immutable and there would be much less confusion?

jasontedor replied:
@rmuir Removed in b50cdd11e27ca6da56bd2a1827e76a0bdb9fdad3

> I dont understand why we have readFrom() at all, vs. say using an abstract class with a ctor that takes Streamable, and thats just how things are deserialized. Then these various classes could be immutable and there would be much less confusion?

I think that we want to go in that direction with all of these eventually.

rmuir commented:
+1
